### DISCUSSION TITLE
Update params return FrozenInstanceError

### DISCUSSION THREAD

**Discussion by marcuswang6** [google/flax#2123] [Q&A] [answered]
Hi,
I got FrozenInstanceError when trying to update TrainState.params with Gradients:

```
state = train_state.TrainState.create(apply_fn=model.apply, params=params, tx=tx)
p = state.params
fn = lambda g, p: (p - lr * gradient)
grads = grad(loss_fn)(p2.params)
p.params = tree_map(fn, grads, p.params)
```

How to fix that error(unfreeze the params)? Or how can I update param p with grads of p2?

Thanks in advance!

**Comment by marcvanzee** [comment, marked as answer]
`FrozenDict` are simply immutable dicts that prevent users from accidentally modifying parameters (which can lead to silent errors). You can unfreeze them with `flax.core.frozen_dict.unfreeze`. We return them from `apply` and `init`.

More generally, we recommend applying gradients through `TrainState`'s [apply_gradients](https://github.com/google/flax/blob/main/flax/training/train_state.py#L58). The easiest way to get started it to take one of our simple examples (e.g., [mnist](https://github.com/google/flax/blob/main/examples/mnist/train.py)) and try modifying the code for your use case. You can also take a look at the [annotated MNIST tutorial](https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html).

**Reply by marcuswang6** [reply]
Thanks for your reply, Marc.
I've tried to unfreeze the params and modify them by following code,  which use grads of p2.params to update p.params:

```
params = unfreeze(model.init(rng, jnp.ones((1024,5,5))))['params']
state = train_state.TrainState.create(apply_fn=model.apply, params=params, tx=tx)
inner_sgd_fn = lambda g, p: (p - outer_lr *g)
grads = grad(loss_fn)(p2.params)
p.params = tree_map(inner_sgd_fn, grads, p.params)#immutable
```

However, it stills return FrozenInstanceError: cannot assign to field 'params'

**Reply by marcvanzee** [reply]
`TrainState` is an immutable pytree, so you can only modify it by making a copy. You can do this using `p = p.replace(params = new_params)`.

**Reply by marcuswang6** [reply]
Now it works, thanks a lot!
